core/ir/instr_shared.c

```diff
@@ -2999,6 +3000,172 @@ instr_uses_fp_reg(instr_t *instr)
+
+    /* Retrieve arithmetic flags from real ISA instruction.
+     * If the real ISA instruction reads or writes one or more arithmetic flag, all
+     * arithmetic flags will be set to read or written in the converted instruction.
+     * Note that this operation can trigger additional encoding and decoding of instr_from
+     * depending on its decoding level.
+     */
+    uint eflags_instr_from = instr_get_arith_flags(instr_from, DR_QUERY_DEFAULT);
+    uint eflags_instr_to = 0;
+    if (TESTANY(EFLAGS_WRITE_ARITH, eflags_instr_from))
+        eflags_instr_to |= EFLAGS_WRITE_ARITH;
+    if (TESTANY(EFLAGS_READ_ARITH, eflags_instr_from))
+        eflags_instr_to |= EFLAGS_READ_ARITH;
+    instr_to->eflags = eflags_instr_to;
+    instr_set_arith_flags_valid(instr_to, true);
+
+    /* Retrieve category of real ISA instruction and set it directly as the category of
+     * the converted instruction.
+     */
+    instr_set_category(instr_to, instr_get_category(instr_from));
+
+    /* Get max_reg_size as opnz_size_t (takes values from OPSZ_ enum), if there are any
+     * operands.
+     */
+    opnd_size_t max_opnd_size = OPSZ_NA;
+    uint num_opnds = num_dsts + num_srcs;
+    if (num_opnds > 0)
+        max_opnd_size = opnd_size_from_bytes(max_reg_size);
+
+    /* Set the source and destination register operands for the converted instruction.
+     * Note that their size is always the same: max_opnd_size.
+     */
+    if (num_dsts > 0) {
+        uint reg_counter = 0;
+        for (uint reg = 0; reg < MAX_NUM_REGS; ++reg) {
+            if (dst_reg_used[reg]) {
+                opnd_t dst_opnd = opnd_create_reg((reg_id_t)reg);
+                opnd_set_size(&dst_opnd, max_opnd_size);
+                instr_set_dst(instr_to, reg_counter, dst_opnd);
+                ++reg_counter;
+            }
+        }
+    }
+
+    if (num_srcs > 0) {
+        uint reg_counter = 0;
+        for (uint reg = 0; reg < MAX_NUM_REGS; ++reg) {
+            if (src_reg_used[reg]) {
+                opnd_t src_opnd = opnd_create_reg((reg_id_t)reg);
+                opnd_set_size(&src_opnd, max_opnd_size);
+                instr_set_src(instr_to, reg_counter, src_opnd);
+                ++reg_counter;
+            }
+        }
+    }
+
+    /* Declare converted instruction operands to be valid.
+     */
+    instr_set_operands_valid(instr_to, true);
+
+    /* Compute instruction length, including bytes for padding to reach 4 byte alignment.
+     * Account for 1 additional byte containing max register operand size, if there are
+     * any operands.
+     */
+    uint num_opnd_bytes = num_opnds > 0 ? num_opnds + 1 : 0;
+    uint instr_length = ALIGN_FORWARD(HEADER_BYTES + num_opnd_bytes, ALIGN_BYTES);
+    instr_to->length = instr_length;
+
+    /* Set converted instruction ISA mode to be DR_ISA_REGDEPS.
+     */
+    instr_set_isa_mode(instr_to, DR_ISA_REGDEPS);
+
+    /* Declare converted instruction valid.
+     */
+    instr_set_raw_bits_valid(instr_to, true);
```

Comment:
There aren't any raw bits.
The reason for instr_set_raw_bits_valid() is to be able to call instr_length() without incurring in additional encoding (instr_length_arch() or private_instr_encode() are called by instr_length() if raw bits are not valid).
If that happens we'd be calling x86 (or arm or whatever) encoding on a REGDEPS instr, and it would break.

However, it is misleading setting to `true` instr_set_raw_bits_valid().
To remove it, there are 2 solutions in my opinion:
1) we don't support calling instr_length() on REGDEPS instructions and we remove it from the supported functions we documented;
2) we support it and add a check for instr_t.isa_mode == DR_ISA_REGDEPS in instr_length().

I went with 1) to avoid spreading more of those checks.

Reply:
> If that happens we'd be calling x86 (or arm or whatever) encoding on a REGDEPS instr, and it would break.

Not following: the ISA is set to regdeps so the encode code will route to the regdeps encoder.

There should be raw bits to avoid re-encoding a decoded instr: which reminds me that the regdeps encoder needs to check for valid raw bits and just do a memcpy if present.

Reply:
> Not following: the ISA is set to regdeps so the encode code will route to the regdeps encoder.
This doesn't happen for REGDEPS instructions because encode_isa_regedeps() is only reachable from instr_encode() and instr_encode_to_copy(), not private_instr_encode().
We did this to avoid "polluting" too many functions with checks for DR_ISA_REGDEPS.
I also noticed that instr_length() calls encode_track_it_block() for ARM, before checking if raw bits are valid.
We probably want to bypass that anyway, so even if we handle the raw bits issue, we'll need to add a check for DR_ISA_REGDEPS in instr_length() if we want to support it.

> There should be raw bits to avoid re-encoding a decoded instr: which reminds me that the regdeps encoder needs to check for valid raw bits and just do a memcpy if present.
Ok, but for REGDEPS instructions who should be responsible to set the raw bits valid and writing to instr_t.bytes?
Should it be instr_convert_to_isa_regdeps()? or decode_isa_regdeps()? or both?
For real ISA instructions I believe private_instr_encode() does it, but as mentioned before I don't think we want to "taint" more functions with the special case:
```
if (isa_mode == DR_REG_DEPS)
  do something else...
else
  do what the original function was doing
```
Also when you say:
> if [raw bits] pointed to the original encoding that is not right
Isn't the original encoding what instr_t.bytes should contain? (a copy of it, to be more precise, but still the original encoding right?)

Reply:
> who should be responsible to set the raw bits valid and writing to instr_t.bytes?

Btw, I'm asking because I don't have a clear use case in mind where we'd want to have instr_t.bytes set.
(except maybe if someone wants to call instr_reuse() on a REGDEPS instructions?)

Reply:
> Ok, but for REGDEPS instructions who should be responsible to set the raw bits valid and writing to instr_t.bytes?

Generally the decoder sets them to point to the original source bits, or when encoding to a private buffer IIRC.

> > if [raw bits] pointed to the original encoding that is not right
> 
> Isn't the original encoding what instr_t.bytes should point to? (a copy of it, to be more precise, but still the original REGDEPS encoding right?)

My comment is in the context of this conversion function from another ISA to regdeps: there, by "original" I meant the other ISA source, so e.g. an x86 encoding's raw bits.

instr_t.bytes points to a valid encoding: it doesn't matter if it's from a decode call or it's a local buffer holding an encoding; if it's set, it's valid (i.e., the IR fields have not changed since that encoding was created/decoded from) and no re-encoding is needed.

Reply:
> This doesn't happen for REGDEPS instructions because encode_isa_regedeps() is only reachable from instr_encode() and instr_encode_to_copy(), not private_instr_encode().
> We did this to avoid "polluting" too many functions with checks for DR_ISA_REGDEPS.

I think we have to refactor those routines to go through a common routine before going to instr_encode_arch(): that is a simple refactor, and it is too fragile to have these commonly-called paths that don't work.

Reply:
> > who should be responsible to set the raw bits valid and writing to instr_t.bytes?
>
> Btw, I'm asking because I don't have a clear use case in mind where we'd want to have instr_t.bytes set.
(except maybe if someone wants to call instr_reuse() on a REGDEPS instruction?)

To avoid repeated encodings: e.g., someone calling instr_length() several times; someone encoding a regdep instr that was just decoded.

Reply:
Understood. Added encoding to bytes field of instr_t in decode_isa_regdeps(), and checking in encode_isa_reg_deps() is raw bits are valid (if so, we memcpy from bytes and return).

Reply:
> I think we have to refactor those routines to go through a common routine before going to instr_encode_arch(): that is a simple refactor, and it is too fragile to have these commonly-called paths that don't work.

Added a common encoding routine: instr_encode_common() (where we check for DR_SIA_REGDEPS and trigger encode_isa_regdeps() is necessary).